### PR TITLE
add layer-TSOS compatibility check in `LayerMeasurements`

### PR DESCRIPTION
**Pull request by missirol**
#### PR description:

This PR includes a possibile solution for the issue discussed in #40174 (see that issue for details), thanks to suggestions from @cms-sw/tracking-pog-l2.

#42318 contains a different solution for the same issue. The two PRs are not mutually exclusive (in principle, both could be integrated).

In this PR, a check is added to `LayerMeasurements` in order to skip `TrajectoryStateOnSurface` objects whose global position is well outside the detector volume. The rationale of this check was discussed in https://github.com/cms-sw/cmssw/issues/40174#issuecomment-1640703758.

#### PR validation:

Manual checks with the reproducers in https://github.com/cms-sw/cmssw/issues/40174#issuecomment-1598702425 and https://github.com/cms-sw/cmssw/issues/40174#issuecomment-1625219782.

#### If this PR is a backport, please specify the original PR and why you need to backport that PR. If this PR will be backported, please specify to which release cycle the backport is meant for:

TBD

### CODE REVIEW

File: TrackingTools/MeasurementDet/src/LayerMeasurements.cc [hunk 1]
```diff
@@ -30,6 +31,26 @@ namespace {
...
+  //      auto const& det = layer.subDetector();
+  //      if (det == PixelBarrel or det == PixelEndcap) {
+  //        return xa < 2e1 and ya < 2e1 and za < 6e1;
+  //      }
+  bool layerIsCompatibleWithTSOS(DetLayer const& /* layer */, TrajectoryStateOnSurface const& tsos) {
+    auto const& gp = tsos.globalPosition();
+    auto const xa = std::abs(gp.x());
+    auto const ya = std::abs(gp.y());
+    auto const za = std::abs(gp.z());
+    return xa < 1e3 and ya < 1e3 and za < 1e3;
```
Comment (missirol) [inline TrackingTools/MeasurementDet/src/LayerMeasurements.cc:51]: These are arguably magic numbers. Suggestions on how to improve this check are welcome.

Comment (mmusich) [reply TrackingTools/MeasurementDet/src/LayerMeasurements.cc:51]: 1e3 cm = 10m ? Should we use instead the inner tracking volume? Or is it used also for muons?

Comment (missirol) [reply TrackingTools/MeasurementDet/src/LayerMeasurements.cc]: Indeed, I removed the detector-type check too quickly. I changed it to 12m in z, which should include the muon detectors.

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mandrenguyen** [comment]
+reconstruction
Fix for HLT crash by requiring a reasonable projected rechit position which is not far away from CMS.
No changes observed in comparisons, as expected.

File: TrackingTools/MeasurementDet/src/LayerMeasurements.cc [hunk 2]
```diff
@@ -30,6 +31,26 @@ namespace {
...
+  //      auto const& det = layer.subDetector();
+  //      if (det == PixelBarrel or det == PixelEndcap) {
+  //        return xa < 2e1 and ya < 2e1 and za < 6e1;
+  //      }
+  bool layerIsCompatibleWithTSOS(DetLayer const& /* layer */, TrajectoryStateOnSurface const& tsos) {
+    auto const& gp = tsos.globalPosition();
+    auto const xa = std::abs(gp.x());
+    auto const ya = std::abs(gp.y());
+    auto const za = std::abs(gp.z());
+    return xa < 1e3 and ya < 1e3 and za < 12e2;
```
Comment (rappoccio) [inline TrackingTools/MeasurementDet/src/LayerMeasurements.cc:51]: Why are these numbers chosen specifically?

Comment (slava77) [reply TrackingTools/MeasurementDet/src/LayerMeasurements.cc:51]: this represents a rough size of CMS

**Comment by missirol** [comment]
Given the changes in data-taking schedule, I guess a backport to `13_0_X` is not so useful anymore (assuming it would have been accepted). I would still backport this to `13_2_X`, unless somebody objects.

**Comment by rappoccio** [comment]
+1

I think it doesn't hurt to have the 13_0_X backport as well, unless there are some conflicts.

**Pull request merged**